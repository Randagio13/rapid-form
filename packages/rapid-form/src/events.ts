type EventHandler = (event: Event) => void;
type EventListenersMap = { [event: string]: EventHandler };

const eventListeners = new WeakMap<Element, EventListenersMap>();

export function addTrackedEventListener(
  element: Element,
  event: string,
  handler: EventHandler
): void {
  if (!eventListeners.has(element)) {
    eventListeners.set(element, {});
  }
  const listeners = eventListeners.get(element)!;
  listeners[event] = handler;
  element.addEventListener(event, handler);
}

export function hasEventListener(element: Element, event: string): boolean {
  return eventListeners.has(element) && !!eventListeners.get(element)?.[event];
}