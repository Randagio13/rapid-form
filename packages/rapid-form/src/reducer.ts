import { type Reducer } from 'react'

/**
 * Represents a value with a name and a corresponding string value.
 */
interface Value {
  name: string
  value: string
}

/**
 * Represents an error associated with a value.
 */
interface Error extends Value {
  isInvalid: boolean
  errorType?: 'required' | 'invalidFormat'
  message?: string
}

/**
 * Represents the state of the reducer.
 */
export interface State {
  values: Record<string, Value>
  errors: Record<string, Error>
}

/**
 * Represents the type of action that can be dispatched to the reducer.
 */
type ActionType = 'setValue' | 'setError' | 'reset'

/**
 * Represents an action that can be dispatched to the reducer.
 */
export interface Action extends State {
  type: ActionType
}

/**
 * Represents the initial state of the reducer.
 */
export const initialState: State = {
  values: {},
  errors: {}
}

/**
 * The reducer function that handles state updates based on dispatched actions.
 * @param state The current state.
 * @param action The action to be dispatched.
 * @returns The new state after applying the action.
 */
export const reducer: Reducer<State, Action> = (state, action) => {
  switch (action.type) {
    case 'setValue':
    case 'setError':
      return { ...state, values: action.values, errors: action.errors }
    case 'reset':
      return { ...state, values: action.values, errors: {} }
  }
}

export default reducer
