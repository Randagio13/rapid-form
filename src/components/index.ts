export { default as Button } from './button/Button'
export { default as Form } from './form/Form'
export { default as Text } from './input/Text'
