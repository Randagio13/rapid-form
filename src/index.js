// @flow
import React from 'react'
import { render } from 'react-dom'
import { AppContainer } from 'react-hot-loader'
import { FormContainer } from 'containers'

const renderApp = (Component) => {
  render((
    <AppContainer>
      <Component />
    </AppContainer>
  ), document.getElementById('app'))
}

renderApp(FormContainer)

if (module.hot) {
  const nextApp = require('./containers').default
  module.hot.accept('./containers', () => { renderApp(nextApp) })
}
