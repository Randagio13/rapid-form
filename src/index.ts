import useRapidForm from './hooks/useRapidForm'

export default useRapidForm
