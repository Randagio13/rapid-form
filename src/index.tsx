import React, { FunctionComponent, useEffect } from 'react'
import * as ReactDOM from 'react-dom'
import useRapidForm from './hooks/useRapidForm'
import { SubmitCallback } from './hooks/useRapidForm'
import _ from 'lodash'
const Form: FunctionComponent = () => {
  const {
    errors,
    validation,
    handleSubmit,
    reset,
    submitValidation,
    values,
    setValue,
    setError,
  } = useRapidForm()
  useEffect(() => {
    console.log('errors', errors)
  }, [errors])
  const s: SubmitCallback = (values, err, e): void => {
    console.log('values', values)
    if (_.isEmpty(err)) {
      reset(e)
    }
  }
  const setDefaultValue = () => {
    setValue('selectData', 'selezione 1')
  }
  const setDefaultError = () => {
    setError({
      name: 'password',
      message: 'Custom error',
      code: 'VALIDATION_ERROR',
    })
  }
  return (
    <form
      id="rapidForm"
      ref={submitValidation}
      autoComplete="off"
      onSubmit={handleSubmit(s)}
    >
      <label>Username:</label>
      <input name="username" ref={validation} required />
      <label id="username-error">{errors.username?.message}</label>
      <br />
      <br />
      <label>Name:</label>
      <input name="name" ref={validation} />
      <br />
      <br />
      <label>Email:</label>
      <input
        name="email"
        type="email"
        ref={validation}
        defaultValue="ale"
        required
      />
      <label id="email-error">{errors.email?.message}</label>
      <br />
      <br />
      <label>Password:</label>
      <input name="password" type="password" ref={validation} required />
      <label id="password-error">{errors.password?.message}</label>
      <br />
      <br />
      <label>Privacy:</label>
      <input name="privacy" type="checkbox" ref={validation} required />
      <label id="privacy-error">{errors.privacy?.message}</label>
      <br />
      <br />
      <label>Calendar:</label>
      <input name="calendar" type="date" ref={validation} required />
      <label id="calendar-error">{errors.calendar?.message}</label>
      <br />
      <br />
      <label>Number:</label>
      <input name="number" type="number" ref={validation} required />
      <label id="number-error">{errors.number?.message}</label>
      <br />
      <br />
      <label>Select:</label>
      <select name="selectData" ref={validation} required defaultValue="">
        <option value="">choose an option</option>
        <option value="selezione 1">selezione 1</option>
        <option value="selezione 2">selezione 2</option>
        <option value="selezione 3">selezione 3</option>
      </select>
      <label id="selectData-error">{errors.selectData?.message}</label>
      <br />
      <br />
      <label>Multi-select:</label>
      <select
        multiple
        name="selectDataMultiple"
        ref={validation}
        required
        defaultValue={['']}
      >
        <option value="">choose an option</option>
        <option value="selezione 1">selezione 1</option>
        <option value="selezione 2">selezione 2</option>
        <option value="selezione 3">selezione 3</option>
      </select>
      <label id="selectDataMultiple-error">
        {errors.selectDataMultiple?.message}
      </label>
      <br />
      <br />
      <label>Radio:</label>
      <input name="radio" type="radio" ref={validation} required value="male" />
      Male
      <br />
      <input name="radio" type="radio" ref={validation} value="female" />
      Female
      <br />
      <input name="radio" type="radio" ref={validation} value="" />
      Nothing
      <label id="radio-error">{errors.radio?.message}</label>
      <br />
      <br />
      Range
      <input
        name="range"
        type="range"
        ref={validation}
        required
        defaultValue="0"
      />
      <label id="radio-error">{errors.range?.message}</label>
      <br />
      <br />
      <button type="button" onClick={setDefaultValue}>
        Set defaultValue
      </button>
      <br />
      <br />
      <button type="button" onClick={setDefaultError}>
        Set Error
      </button>
      <br />
      <br />
      <pre>{`values: ${JSON.stringify(values, null, 2)}`}</pre>
      <pre>{`errors: ${JSON.stringify(errors, null, 2)}`}</pre>
    </form>
  )
}

ReactDOM.render(<Form />, document.getElementById('root'))
