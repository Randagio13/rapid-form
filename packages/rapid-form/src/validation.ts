import { type Dispatch } from 'react'
import { type Action } from 'reducer.js'

type EventType = 'change' | 'blur' | 'input'

/**
 * Validation function
 * return true if the value is valid
 */
type ValidationFunction = (props: {
  value: string
  formElements: HTMLFormControlsCollection
}) => boolean

type ValidationConfig = Record<
  string,
  {
    validation: ValidationFunction
    eventType?: EventType
    message?: string
  }
>

export interface ValidationProps {
  /**
   * Reference to form element
   */
  ref: HTMLFormElement | null
  /**
   * Dispatch action to reducer
   */
  dispatch?: Dispatch<Action>
  /**
   * Configure event type for validation (change, blur, input)
   * @default { eventType: 'input', resetOnSubmit: true }
   */
  config?:
    | {
        eventType?: EventType
        validations?: ValidationConfig
        resetOnSubmit?: boolean
      }
    | undefined
}

/**
 * Validate email address using regex
 * @param email - The email address to validate
 * @returns True if the email address is valid, false otherwise
 */
function validateEmail(email: string): boolean {
  const re =
    // eslint-disable-next-line no-useless-escape
    /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  return re.test(String(email).toLowerCase())
}

interface InputValidationProps {
  /**
   * Input type
   */
  type: HTMLInputElement['type']
  /**
   * Input value
   */
  value: string
  /**
   * Current element
   */
  element: HTMLInputElement
}

/**
 * Validate input value based on type
 * @param props - The input validation properties
 * @returns True if the input value is valid, false otherwise
 */
function inputValidation({ type, value, element }: InputValidationProps): boolean {
  switch (type) {
    case 'email':
      return validateEmail(value)
    case 'password':
      return value.length > 6
    case 'checkbox':
      return element.checked
    default:
      return value.length > 0
  }
}

/**
 * Perform validation on form elements
 * @param props - The validation properties
 */
export function validation({ ref, dispatch, config }: ValidationProps): void {
  const elements = ref?.elements
  let eventType = config?.eventType ?? 'input'
  const resetOnSubmit = config?.resetOnSubmit ?? true
  if (resetOnSubmit) {
    ref?.addEventListener('submit', function () {
      ref?.reset()
      dispatch?.({
        type: 'reset',
        values: {},
        errors: {}
      })
    })
  }
  if (elements != null) {
    for (let i = 0; i < elements.length; i++) {
      const element = elements[i]
      if (element != null) {
        const isRequired = element?.hasAttribute('required')
        if (isRequired) {
          const currentElementName = element.getAttribute('name') as string
          eventType =
            config?.validations?.[currentElementName]?.eventType ?? eventType
          element?.addEventListener(eventType, function (e) {
            const target = e.target as HTMLInputElement
            const val = target.value.trim()
            const isValid =
              config?.validations?.[target.name]?.validation != null
                ? config?.validations[target.name]?.validation({
                    value: val,
                    formElements: elements
                  })
                : inputValidation({
                    type: target.type,
                    value: val,
                    element: target
                  })
            if (isValid === true) {
              dispatch?.({
                type: 'setValue',
                values: { [target.name]: { name: target.name, value: val } },
                errors: {
                  [target.name]: {
                    name: target.name,
                    value: val,
                    isInvalid: false
                  }
                }
              })
            } else {
              dispatch?.({
                type: 'setError',
                values: { [target.name]: { name: target.name, value: val } },
                errors: {
                  [target.name]: {
                    name: target.name,
                    value: val,
                    isInvalid: true,
                    errorType: 'invalidFormat',
                    message:
                      config?.validations?.[target.name]?.message ??
                      'Invalid format or required field'
                  }
                }
              })
            }
          })
        }
      }
    }
  }
}
