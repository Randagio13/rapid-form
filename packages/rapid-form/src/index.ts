import { useReducer } from 'react'
import reducer, { type State, initialState } from 'reducer.js'
import { type ValidationProps, validation } from 'validation.js'

export type Config = ValidationProps['config']
export type Value = State['values'][string]
export type Error = State['errors'][string]

/**
 * Rapid form hook
 */
export function useRapidForm(): {
  refValidation: (
    ref: HTMLFormElement | null,
    config?: ValidationProps['config']
  ) => void
  values: State['values']
  errors: State['errors']
} {
  const [state, dispatch] = useReducer(reducer, initialState)
  return {
    refValidation: (ref, config) => {
      validation({ ref, dispatch, config, state })
    },
    ...state
  }
}
