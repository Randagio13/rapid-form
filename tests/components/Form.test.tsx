import { shallow } from 'enzyme'
import * as React from 'react'
import { create } from 'react-test-renderer'
import { Form } from '../../src/components'

export const testOnSubmit = (event: any) => {
  alert(event)
}

const component = create(
  <Form id='formId' onSubmit={testOnSubmit} disabled={true}><input type='text' /></Form>
).toJSON()

test('Snapshot', () => {
  expect(component).toMatchSnapshot()
})

test('Check property component', () => {
  expect(component.type).toEqual('form')
  expect(component.props.id).toEqual('formId')
  expect(component.props.name).toEqual('formId')
  expect(component.props.disabled).toEqual(true)
})

test('Check correct name if there is prop', () => {
  const checkName = create(
    <Form id='formId' name='formName' onSubmit={testOnSubmit}><input type='text' /></Form>
  ).toJSON()
  expect(checkName.props.name).toEqual('formName')
  expect(checkName.props.name).not.toEqual('formId')
})
