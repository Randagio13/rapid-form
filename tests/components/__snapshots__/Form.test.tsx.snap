// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<form
  disabled={true}
  id="formId"
  name="formId"
>
  <input
    type="text"
  />
</form>
`;
